azure-iot-device/azure/iot/device/iothub/abstract_clients.py

```diff
@@ -67,6 +68,24 @@ def _form_sas_uri(hostname, device_id, module_id=None):
         return "{hostname}/devices/{device_id}".format(hostname=hostname, device_id=device_id)
 
 
+def _extract_sas_uri_values(uri):
```

Comment:
>_extract_sas_uri_values [](start = 4, length = 23)

pedantic comment: this is parsing a _resource_, not a uri.  While verifying this, I found more info on sas tokens than I ever wanted to know: [https://docs.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas](https://docs.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas)

Reply:
And over here they say it's a URI: [https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security)

MS docs on SAStoken are NOT GOOD

This document you've provided does seem to have a LOT more info than any one I've seen before. Would be really nice if someone would just collect all the SAS token information, and put it in one place, and it was just always kept up to date and accurate forever. And also explained things, like what an auth rule is, and why it's called that, when it's clearly a SharedKeyName, thus the "skn" variable.

---
In reply to: [506685317](https://github.com/Azure/azure-iot-sdk-python/pull/692#discussion_r506685317) [](ancestors = 506685317)